lib/connection/connect.js

```diff
@@ -25,13 +28,106 @@ function connect(options, callback) {
           return;
         }
 
-        callback(null, new Connection(ipv4Socket, options));
+        performInitialHandshake(new Connection(ipv4Socket, options), options, callback);
       });
 
       return;
     }
 
-    callback(null, new Connection(ipv6Socket, options));
+    performInitialHandshake(new Connection(ipv6Socket, options), options, callback);
+  });
+}
+
+function isSupportedServer(ismaster) {
+  return ismaster && typeof ismaster.maxWireVersion === 'number' && ismaster.maxWireVersion >= 2;
+}
+
+function getSaslSupportedMechs(options) {
+  if (!(options && options.credentials)) {
+    return {};
+  }
+
+  const credentials = options.credentials;
+  const authMechanism = credentials.mechanism;
+  const authSource = credentials.source || options.dbName || 'admin';
+  const user = credentials.username || options.user;
+
+  if (typeof authMechanism === 'string' && authMechanism.toUpperCase() !== 'DEFAULT') {
+    return {};
+  }
+
+  if (!user) {
+    return {};
+  }
+
+  return { saslSupportedMechs: `${authSource}.${user}` };
+}
+
+function performInitialHandshake(conn, options, callback) {
+  let compressors = [];
+  if (options.compression && options.compression.compressors) {
+    compressors = options.compression.compressors;
+  }
+
+  const handshakeDoc = Object.assign(
+    {
+      ismaster: true,
+      client: createClientInfo(options),
+      compression: compressors
+    },
+    getSaslSupportedMechs(options)
+  );
+
+  const start = new Date().getTime();
+  runCommand(conn, 'admin.$cmd', handshakeDoc, options, (err, ismaster) => {
+    if (err) {
+      callback(err, null);
+      return;
+    }
+
+    if (ismaster.ok === 0) {
+      callback(new MongoError(ismaster), null);
+      return;
+    }
+
+    if (!isSupportedServer(ismaster)) {
+      const latestSupportedVersion = '2.6';
+      const message =
+        'Server at ' +
+        options.host +
+        ':' +
+        options.port +
+        ' reports wire version ' +
+        (ismaster.maxWireVersion || 0) +
+        ', but this version of Node.js Driver requires at least 2 (MongoDB' +
```

Comment:
I'm a little confused by the wording of this error message...should it be `but this version of Node.js Driver requires at least MongoDB + latestSupportedVersion + ').'`? Why does it say `requires at least 2 (MongoDB`?

Reply:
This is supposed to be configurable (see new SDAM implementation), but what it's saying is that the wire version reported from the remote server is not at least the value `2` (MongoDB 2.6). `latestSupportedVersion` is the MongoDB version, what we would need here is the `latestSupportedMaxWireVersion`

Reply:
added `latestSupportedMaxWireVersion`